### ISSUE TITLE
How did flash-attn compute attention for cu_seqlens

### ISSUE THREAD

**Issue by zigzagcai** [Dao-AILab/flash-attention#850]
Hi,

We know that `cu_seqlens` is for the compute efficiency when we do training over multiple variable-length samples. And the attention mask can be calculated through `cu_seqlens`. We can cut the original cumulative sequence into a batch sequence and pad the empty positions with zeros. But this approach will waste training efficiency, since computing resources are consumed for meaningless padding tokens.

I am not quite familiar with the implementation details of flash-attn. So, I am just curious about where can I find the implementation or mechanism that how did flash-attn compute attention directly over cumulative sequence and get the separate results?

Thanks!

**Comment by tridao** [comment]
We just launch parallel work (i.e. thread blocks) to process each attn head of each sequence, and each thread block will figure out the start and end idx of each sequence from cu_seqlens.

**Comment by zigzagcai** [comment]
> We just launch parallel work (i.e. thread blocks) to process each attn head of each sequence, and each thread block will figure out the start and end idx of each sequence from cu_seqlens.

Got it. Thanks for the explanation!

**Comment by zigzagcai** [comment]
> We just launch parallel work (i.e. thread blocks) to process each attn head of each sequence, and each thread block will figure out the start and end idx of each sequence from cu_seqlens.

We observe that the Flash API provides fwd/bwd and varlen_fwd/varlen_bwd to handle inputs both with and without cu_seqlen. Different inputs all proceed to run_mha_fwd and run_mha_bwd, and ultimately are processed by the kernel within the Flash attention mechanism.

In the kernel of Flash attention, a structure named blockinfo is defined to store the offsets of Q, K, and V (qkv). These offsets are calculated based on cu_seqlen_q and cu_seqlen_k, allowing for the computation of attention on a per-row basis. Moreover, each row's qkv supports variable lengths, thereby preventing the waste of computational resources on padding tokens.

**Comment by zigzagcai** [comment]
The above is my understanding after looking at the source cuda code of flash attention.
I’m not sure if my understanding is correct?

**Comment by 2hiTee** [comment]
> Hi,
> 
> We know that `cu_seqlens` is for the compute efficiency when we do training over multiple variable-length samples. And the attention mask can be calculated through `cu_seqlens`. We can cut the original cumulative sequence into a batch sequence and pad the empty positions with zeros. But this approach will waste training efficiency, since computing resources are consumed for meaningless padding tokens.
> 
> I am not quite familiar with the implementation details of flash-attn. So, I am just curious about where can I find the implementation or mechanism that how did flash-attn compute attention directly over cumulative sequence and get the separate results?
> 
> Thanks!

Hi, you mentioned that original code for calculating attention output needs attention_mask, which can be calculated using cu_seqlens. So I am wondering how to convert cu_seqlens to attention_mask, I would be glad if you could give some advice. Thanks!
